### ISSUE TITLE
Docker volume mapping not working

### ISSUE THREAD

**Issue by zmorris** [docker/for-mac#2991]
- [x] I have tried with the latest version of my channel (Stable)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 2E751024-8F4B-4F1C-8DDC-073D0F65D2A0/20180611-153130

### Expected behavior

Chrome should show `Hello world`

### Actual behavior

Chrome shows `This site can’t be reached` and `docker ps` no longer shows the container.

### Information

  - macOS Version: High Sierra 10.13.4 (17E199)

### Diagnostic logs

All were `ok`, I accidentally closed the window before this section.

```
Docker for Mac: Version 18.03.1-ce-mac65 (24312)
```

### Steps to reproduce the behavior

I'm using a fresh install of Docker for Mac.  I submitted a full example at:

Stack Overflow: https://stackoverflow.com/questions/50805885/docker-volume-mapping-not-working
Forums: https://forums.docker.com/t/docker-volume-mapping-not-working/52236

1. Download the zip https://repl.it/repls/BurlyAncientTrust.zip
1. Unzip it
1. `cd` into it
1. run:

    ```
    (image=node-web-app && docker build -t $image . && docker run --rm -p 49160:8080 -d $image)
    ```

1. Visit http://localhost:49160/ in Chrome and see `Hello world`
1. run with the `-v` option:

    ```
    (image=node-web-app && docker ps -a -q --filter=ancestor=$image | xargs docker rm -f && docker rmi $image)
    (image=node-web-app && docker build -t $image . && docker run --rm -v "$(pwd)":/usr/src/app -p 49160:8080 -d $image)
    ```

1. Visit http://localhost:49160/ in Chrome and see `This site can’t be reached`
1. run:

    ```
    docker ps
    ```

1.  `node-web-app` container is no longer shown

**Comment by djs55** [comment]
If you run it without the `-d` option (which means "run in the background") then it prints:
```
$ docker run --rm -v "$(pwd)":/usr/src/app -p 49160:8080 node-web-app

> docker_web_app@1.0.0 start /usr/src/app
> node index.js

module.js:549
    throw err;
    ^

Error: Cannot find module 'express'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/usr/src/app/index.js:3:17)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! docker_web_app@1.0.0 start: `node index.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the docker_web_app@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2018-06-12T12_08_54_394Z-debug.log
```
This explains why `docker ps` doesn't show it -- the error made the app quit.

If you look inside the image with:
```
$ docker run -it node-web-app bash
root@6d382985259d:/usr/src/app# ls
Dockerfile  index.js  node_modules  package-lock.json  package.json
```
the image has `/usr/src/app/node_modules` directory created by the `RUN npm install` step in the `Dockerfile`. When the host's directory is mounted on top of `/usr/src/app`, the `node_modules` directory is hidden.

I think you should investigate moving the `index.js` into a subdirectory and bind mount over that (e.g. `/usr/src/app/code`) so that the `node_modules` directory is left alone, or investigate moving the `node_modules` directory somewhere else.

Hope that helps!

I'll close this issue as I think Docker is working as expected.

**Comment by zmorris** [comment]
Thank you very much for your quick reply!  I'm just learning Docker so the various running modes and verbose logging are still new to me.  I had a hunch that maybe the directory sync from host to container stomped the files inside the container but had no idea where to begin.

Ok I bit the bullet and dug deeper on this.  As I understand it, this happens currently:

1. Dockerfile calls `RUN npm install`
1. `-v "$(pwd)":/usr/src/app` writes the contents of host `"$(pwd)"` over container `/usr/src/app`
1. Node.js no longer sees `/usr/src/app/node_modules` so crashes and the container stops

But I was under the impression that Docker volumes were overlayed:

https://docs.docker.com/storage/storagedriver/overlayfs-driver/#how-the-overlay-driver-works

So container files take priority over image files.  In this case the container is run and makes the contents of `"$(pwd)"` into a volume `/usr/src/app` in the container.

What I don't understand is why image `node_modules` doesn't pass through and remain visible as container `node_modules`.  In other words, why isn't container `/usr/src/app` overlayed above image `/usr/src/app`?  Instead, `/usr/src/app` gets replaced, which seems... odd.

All I could find about that is https://github.com/moby/moby/issues/10947#issuecomment-76652694 from cpuguy83:
```
Hi, I'm not sure I understand. Do you literally mean overlay (as in merge) a host path with a container path such as what might be done with overlay fs?

Docker does not support this directly, however you can create an overlay on the host and then mount that in.

Volumes do automatically copy data from the image to the host for "normal" volumes, ie volumes like docker run -v /foo, where as volumes created as such docker run -v /host/path:/container/path will not.
```

This may be a **conceptual bug** from the early history of Docker.  To me, there should be no distinction between `docker run -v /foo` and `run -v /host/path:/container/path`.  Does someone have a satisfactory answer for why this behavior is inconsistent?

I think as a workaround, what I can maybe do is mount the host directory in the container as a volume.  Then from within the container, overlay another volume on top of that at `/usr/src/app`, with its contents preserved.  That way the container can continue to write things like `/usr/src/app/node_modules` and if I edit any files on the host, those changes will show up in the container.  https://stackoverflow.com/a/30101039/539149 comes pretty close to that.

Another fix to all of this would be for `docker run -v "$(pwd)":/usr/src/app` to create the mount and map it before the `Dockerfile` is run.  This may be impossible though, because the Dockerfile probably needs to create the container first.

Another wrench in all of this is that in order for Dockerfile to be portable, a decision was made early on to not mount host volumes during build https://stackoverflow.com/a/26053710/539149 which I agree with, but could be the underlying cause of this situation.

Please, if someone stumbles onto this who has been through this before, can you reply with a magical incantation that runs a docker container with the directory mapped from the host?  This is literally one of the very first things I tried with Docker, and since it didn't work, I'm not feeling super confident about what lies ahead.

TL;DR: I want to sync a directory from the host to one in the container, without side effects like the ones discussed above.

**Comment by zmorris** [comment]
I found a partial workaround and wrote up a solution at https://stackoverflow.com/a/50825328/539149

The gist of it is that `node_modules` can be mapped as an additional volume with `-v /usr/src/app/node_modules`.

I'm thinking this works because it makes `node_modules` an overlayed volume, which preserves any files inside it rather than overwriting them.

It's not ideal because it still requires dealing with the contents of the image (specifically the newly-added `node_modules` directory).  If someone knows a better way, please post it thanks!
